src/it/scala/com/wavesplatform/it/BalancesSpecification.scala

```diff
@@ -0,0 +1,299 @@
+    } yield succeed
+
+    Await.result(f, 1 minute)
+  }
+
+  test("сan not make leasing without having enough waves") {
+    val f = for {
+      fb <- Future.traverse(allNodes)(_.height).map(_.min)
+
+      _ <- assertBalances(firstAddress, 190 waves, 90 waves)
+      _ <- assertBalances(secondAddress, 100 waves, 200 waves)
+
+      leaseFailureAssertion <- recoverToSucceededIf[IOException] {
+        sender.lease(secondAddress, firstAddress, 101 waves, 10 waves)
+      }
+
+      _ <- Future.traverse(allNodes)(_.waitForHeight(fb + 2))
+
+      _ <- assertBalances(firstAddress, 190 waves, 90 waves)
+      _ <- assertBalances(secondAddress, 100 waves, 200 waves)
+    } yield leaseFailureAssertion
+
+    Await.result(f, 1 minute)
+  }
+
+  test("can not make transfer without having enough of your own waves") {
+    val f = for {
+      fb <- Future.traverse(allNodes)(_.height).map(_.min)
+
+      _ <- assertBalances(firstAddress, 190 waves, 90 waves)
+      _ <- assertBalances(secondAddress, 100 waves, 200 waves)
+
+      transferFailureAssertion <- recoverToSucceededIf[IOException] {
+        sender.transfer(secondAddress, firstAddress, 101 waves, fee = 1 waves)
+      }
+
+      _ <- Future.traverse(allNodes)(_.waitForHeight(fb + 2))
+
+      _ <- assertBalances(firstAddress, 190 waves, 90 waves)
+      _ <- assertBalances(secondAddress, 100 waves, 200 waves)
+    } yield transferFailureAssertion
+
+    Await.result(f, 1 minute)
+  }
+
+  test("can not make transfer without having enough of effective balance") {
+    val f = for {
+      fb <- Future.traverse(allNodes)(_.height).map(_.min)
+
+      _ <- assertBalances(firstAddress, 190 waves, 90 waves)
+      _ <- assertBalances(secondAddress, 100 waves, 200 waves)
+
+      transferFailureAssertion <- recoverToSucceededIf[IOException] {
+        sender.transfer(firstAddress, secondAddress, 91 waves, fee = 1 waves)
+      }
+
+      _ <- Future.traverse(allNodes)(_.waitForHeight(fb + 2))
+
+      _ <- assertBalances(firstAddress, 190 waves, 90 waves)
+      _ <- assertBalances(secondAddress, 100 waves, 200 waves)
+    } yield transferFailureAssertion
+
+    Await.result(f, 1 minute)
+  }
+
+  test("can not make leasing without having enough waves for fee") {
+    val f = for {
+      fb <- Future.traverse(allNodes)(_.height).map(_.min)
+
+      _ <- assertBalances(firstAddress, 190 waves, 90 waves)
+      _ <- assertBalances(secondAddress, 100 waves, 200 waves)
+
+      transferFailureAssertion <- recoverToSucceededIf[IOException] {
+        sender.lease(firstAddress, secondAddress, 88 waves, fee = 3 waves)
+      }
+
+      _ <- Future.traverse(allNodes)(_.waitForHeight(fb + 2))
+
+      _ <- assertBalances(firstAddress, 190 waves, 90 waves)
+      _ <- assertBalances(secondAddress, 100 waves, 200 waves)
+    } yield transferFailureAssertion
+
+    Await.result(f, 1 minute)
+  }
+
+
+  test("leasing cancel is working correctly") {
```

Comment:
"lease cancellation reverts eff.b. changes; lessor pays fee for both lease and cancellation"